File: easy/0401_binary_watch/solution.cpp
#include <string>
#include <vector>

class Solution {
public:
  std::vector<std::string> readBinaryWatch(int turnedOn) {
    if (turnedOn == 0)
      return std::vector<std::string>{"0:00"};
    if (turnedOn > 8)
      return std::vector<std::string>{};

    int hour, minute;

    std::vector<std::pair<int, std::string>> hours;
    hours.reserve(12);
    for (hour = 0; hour < 12; ++hour)
      hours.push_back(std::pair<int, std::string>{__builtin_popcount(hour),
                                                  std::to_string(hour)});

    std::vector<std::pair<int, std::string>> minutes;
    minutes.reserve(60);
    for (minute = 0; minute < 60; ++minute)
      minutes.push_back(std::pair<int, std::string>{
          __builtin_popcount(minute),
          minute < 10 ? '0' + std::to_string(minute) : std::to_string(minute)});

    std::vector<std::string> times;

    for (hour = 0; hour < 12; ++hour) {
      if (hours[hour].first <= turnedOn) {
        for (minute = 0; minute < 60; ++minute) {
          if (hours[hour].first + minutes[minute].first == turnedOn)
            times.push_back(hours[hour].second + ':' + minutes[minute].second);
        }
      }
    }

    return times;
  }
};

File: easy/0401_binary_watch/solution.py
from typing import List


class Solution:
    def readBinaryWatch(self, turned_on: int) -> List[str]:
        if turned_on == 0:
            return ["0:00"]
        if turned_on > 8:
            return []

        times = []

        for hour in range(12):
            if hour.bit_count() <= turned_on:
                time = str(hour)

                for minutes in range(60):
                    if minutes.bit_count() == turned_on - hour.bit_count():
                        times.append(f"{time}:{minutes:02d}")

        return times
